File: Registro.pl
main :-
    cargar_datos(Estudiantes),
    menu(Estudiantes).

cargar_datos(Estudiantes) :-
    open('UniversityPL.txt', read, Stream),
    read(Stream, Estudiantes),
    close(Stream).

guardar_datos(Estudiantes) :-
    open('UniversityPL.txt', write, Stream),
    write(Stream, Estudiantes),
    write(Stream, '.'),
    close(Stream).

menu(Estudiantes) :-
    nl, write('--- SISTEMA DE REGISTRO UNIVERSIDAD ---'), nl,
    write('1) Check In'), nl,
    write('2) Buscar por ID'), nl,
    write('3) Calculo de Tiempo'), nl,
    write('4) Listar Estudiantes'), nl,
    write('5) Check Out'), nl,
    write('0) Salir'), nl,
    write('Elige una opcion (termina con punto): '), nl,
    read(Opcion),
    procesar_opcion(Opcion, Estudiantes).

procesar_opcion(1, Estudiantes) :-
    nl, write('Ingrese ID del estudiante (con comillas simples, ej. ''1001''.): '), nl,
    read(ID),
    ( number(ID) ->
        write('-> Error: El ID debe ingresarse como texto (entre comillas simples).'), nl,
        menu(Estudiantes)
    ;
        ( buscar_estudiante(ID, Estudiantes, _) ->
            write('-> Error: El estudiante ya se encuentra registrado.'), nl,
            menu(Estudiantes)
        ;
            write('Ingrese hora de entrada (minutos desde 00:00, solo enteros, ej. 480.): '), nl,
            read(Hora),
            ( integer(Hora) ->
                NuevoEstudiante = estudiante(ID, Hora),
                NuevosEstudiantes = [NuevoEstudiante | Estudiantes],
                guardar_datos(NuevosEstudiantes),
                write('-> Estudiante registrado exitosamente.'), nl,
                menu(NuevosEstudiantes)
            ;
                write('-> Error: La hora de entrada debe ser un numero entero.'), nl,
                menu(Estudiantes)
            )
        )
    ).

procesar_opcion(2, Estudiantes) :-
    nl, write('Ingrese ID a buscar (ej. ''1001''.): '), nl,
    read(ID),
    ( buscar_estudiante(ID, Estudiantes, Hora) ->
        write('-> Estudiante encontrado. Entro a los '), write(Hora), write(' minutos.'), nl
    ;
        write('-> Estudiante no encontrado o ya salio.'), nl
    ),
    menu(Estudiantes).

procesar_opcion(3, Estudiantes) :-
    nl, write('Ingrese ID del estudiante (ej. ''1001''.): '), nl,
    read(ID),
    ( buscar_estudiante(ID, Estudiantes, HEntrada) ->
        write('Ingrese hora actual (minutos desde 00:00, ej. 540.): '), nl,
        read(HSalida),
        ( integer(HSalida), HSalida >= HEntrada ->
            Duracion is HSalida - HEntrada,
            Horas is Duracion // 60,
            Minutos is Duracion mod 60,
            write('-> Tiempo de permanencia: '), write(Horas), write(' horas y '), write(Minutos), write(' minutos.'), nl
        ;
            write('-> Error: La hora actual debe ser un numero entero y no puede ser anterior a la de entrada.'), nl
        )
    ;
        write('-> Estudiante no encontrado.'), nl
    ),
    menu(Estudiantes).

procesar_opcion(4, []) :-
    nl, write('--- Estudiantes Registrados ---'), nl,
    write('-> No hay ningun estudiante registrado en este momento.'), nl,
    menu([]).

procesar_opcion(4, Estudiantes) :-
    Estudiantes \= [],
    nl, write('--- Estudiantes Registrados ---'), nl,
    write(Estudiantes), nl,
    menu(Estudiantes).

procesar_opcion(5, Estudiantes) :-
    nl, write('Ingrese ID a dar salida (ej. ''1001''.): '), nl,
    read(ID),
    ( buscar_estudiante(ID, Estudiantes, HEntrada) ->
        write('Ingrese hora de salida (minutos desde 00:00, ej. 600.): '), nl,
        read(HSalida),
        ( integer(HSalida), HSalida >= HEntrada ->
            Duracion is HSalida - HEntrada,
            eliminar_estudiante(ID, Estudiantes, NuevosEstudiantes),
            guardar_datos(NuevosEstudiantes),
            write('-> Check Out exitoso. Tiempo total: '), write(Duracion), write(' minutos.'), nl,
            menu(NuevosEstudiantes)
        ;
            write('-> Error: La hora de salida debe ser un entero y no puede ser anterior a la de entrada.'), nl,
            menu(Estudiantes)
        )
    ;
        write('-> Estudiante no encontrado.'), nl,
        menu(Estudiantes)
    ).

procesar_opcion(0, _) :-
    nl, write('Saliendo del sistema...'), nl.

procesar_opcion(_, Estudiantes) :-
    nl, write('Opcion invalida. Intenta de nuevo.'), nl,
    menu(Estudiantes).

buscar_estudiante(ID, [estudiante(ID, Hora) | _], Hora).
buscar_estudiante(ID, [estudiante(OtroID, _) | Cola], Hora) :-
    ID \= OtroID,
    buscar_estudiante(ID, Cola, Hora).

eliminar_estudiante(_, [], []).
eliminar_estudiante(ID, [estudiante(ID, _) | Cola], Cola).
eliminar_estudiante(ID, [estudiante(OtroID, Hora) | Cola], [estudiante(OtroID, Hora) | NuevaCola]) :-
    ID \= OtroID,
    eliminar_estudiante(ID, Cola, NuevaCola).

File: Registro.hs
import System.IO
import Text.Read (readMaybe)

type Estudiante = (String, Int)
type Universidad = [Estudiante]

archivo :: String
archivo = "University.txt"

main :: IO ()
main = do
    contenido <- readFile archivo
    let estudiantes = length contenido `seq` read contenido :: Universidad
    menu estudiantes

menu :: Universidad -> IO ()
menu ests = do
    putStrLn "\n--- SISTEMA DE REGISTRO UNIVERSIDAD ---"
    putStrLn "1) Check In"
    putStrLn "2) Buscar por ID"
    putStrLn "3) Calculo de Tiempo"
    putStrLn "4) Listar Estudiantes"
    putStrLn "5) Check Out"
    putStrLn "0) Salir"
    putStr "Elige una opcion: "
    opcion <- getLine
    
    case opcion of
        "1" -> checkIn ests
        "2" -> buscar ests
        "3" -> calcularTiempo ests
        "4" -> listar ests
        "5" -> checkOut ests
        "0" -> putStrLn "\nSaliendo del sistema..."
        _   -> putStrLn "\nOpcion invalida." >> menu ests

checkIn :: Universidad -> IO ()
checkIn ests = do
    putStr "Ingrese ID del estudiante: "
    idEst <- getLine
    
    case buscarEstudiante idEst ests of
        Just _  -> do
            putStrLn "-> Error: El estudiante ya se encuentra registrado en el sistema."
            menu ests
        Nothing -> do
            putStr "Ingrese hora de entrada (minutos desde 00:00, ej. 480): "
            horaStr <- getLine
            
            case readMaybe horaStr :: Maybe Int of
                Just hora -> do
                    let nuevosEsts = (idEst, hora) : ests
                    guardar nuevosEsts
                    putStrLn "-> Estudiante registrado exitosamente."
                    menu nuevosEsts
                Nothing -> do
                    putStrLn "-> Error: Formato de hora invalido. Por favor, ingresa un numero entero."
                    menu ests

buscar :: Universidad -> IO ()
buscar ests = do
    putStr "Ingrese ID a buscar: "
    idEst <- getLine
    case buscarEstudiante idEst ests of
        Just (i, h) -> putStrLn $ "-> Estudiante " ++ i ++ " esta en la universidad. Entro a los " ++ show h ++ " minutos."
        Nothing     -> putStrLn "-> Estudiante no encontrado o ya salio."
    menu ests

calcularTiempo :: Universidad -> IO ()
calcularTiempo ests = do
    putStr "Ingrese ID del estudiante: "
    idEst <- getLine
    case buscarEstudiante idEst ests of
        Just (_, hEntrada) -> do
            putStr "Ingrese hora actual (minutos desde 00:00): "
            hSalidaStr <- getLine
            
            case readMaybe hSalidaStr :: Maybe Int of
                Just hSalida -> 
                    if hSalida >= hEntrada 
                        then do
                            let duracion = hSalida - hEntrada
                                horas = duracion `div` 60
                                minutos = duracion `mod` 60
                            putStrLn $ "-> Tiempo de permanencia: " ++ show horas ++ " horas y " ++ show minutos ++ " minutos."
                        else putStrLn "-> Error: La hora actual no puede ser anterior a la hora de entrada."
                Nothing -> putStrLn "-> Error: Formato invalido. Ingresa un numero entero."
        Nothing -> putStrLn "-> Estudiante no encontrado."
    menu ests

listar :: Universidad -> IO ()
listar ests = do
    putStrLn "\n--- Estudiantes Registrados ---"
    if null ests
        then putStrLn "-> No hay ningun estudiante registrado en este momento."
        else print ests
    menu ests

checkOut :: Universidad -> IO ()
checkOut ests = do
    putStr "Ingrese ID del estudiante a dar salida: "
    idEst <- getLine
    case buscarEstudiante idEst ests of
        Just (_, hEntrada) -> do
            putStr "Ingrese hora de salida (minutos desde 00:00): "
            hSalidaStr <- getLine
            
            case readMaybe hSalidaStr :: Maybe Int of
                Just hSalida -> 
                    if hSalida >= hEntrada
                        then do
                            let duracion = hSalida - hEntrada
                                nuevosEsts = eliminarEstudiante idEst ests
                            guardar nuevosEsts
                            putStrLn $ "-> Check Out exitoso. Tiempo total: " ++ show duracion ++ " minutos."
                            menu nuevosEsts
                        else do
                            putStrLn "-> Error: La hora de salida no puede ser anterior a la hora de entrada."
                            menu ests
                Nothing -> do
                    putStrLn "-> Error: Formato invalido. Ingresa un numero entero."
                    menu ests
        Nothing -> do
            putStrLn "-> Estudiante no encontrado."
            menu ests

guardar :: Universidad -> IO ()
guardar ests = writeFile archivo (show ests)

buscarEstudiante :: String -> Universidad -> Maybe Estudiante
buscarEstudiante _ [] = Nothing
buscarEstudiante idBuscado ((i, h):resto)
    | idBuscado == i = Just (i, h)
    | otherwise      = buscarEstudiante idBuscado resto

eliminarEstudiante :: String -> Universidad -> Universidad
eliminarEstudiante _ [] = []
eliminarEstudiante idBuscado ((i, h):resto)
    | idBuscado == i = resto
    | otherwise      = (i, h) : eliminarEstudiante idBuscado resto